packages/core/src/transports/base.ts

```diff
@@ -0,0 +1,147 @@
+  bufferSize?: number;
+}
+
+export interface BaseTransportOptions extends InternalBaseTransportOptions {
+  // url to send the event
+  // transport does not care about dsn specific - client should take care of
+  // parsing and figuring that out
+  url: string;
+}
+
+// TODO: Move into Browser Transport
+export interface BrowserTransportOptions extends BaseTransportOptions {
+  // options to pass into fetch request
+  fetchParams: Record<string, string>;
+  headers?: Record<string, string>;
+  sendClientReports?: boolean;
+}
+
+// TODO: Move into Node transport
+export interface NodeTransportOptions extends BaseTransportOptions {
+  // Set a HTTP proxy that should be used for outbound requests.
+  httpProxy?: string;
+  // Set a HTTPS proxy that should be used for outbound requests.
+  httpsProxy?: string;
+  // HTTPS proxy certificates path
+  caCerts?: string;
+}
+
+interface NewTransport {
+  // If `$` is set, we know that this is a new transport.
+  // TODO(v7): Remove this as we will no longer have split between
+  // old and new transports.
+  $: boolean;
+  send(request: Envelope, category: TransportCategory): PromiseLike<TransportResponse>;
+  flush(timeout?: number): PromiseLike<boolean>;
+}
+
+type MakeTransportRequest = (request: TransportRequest) => PromiseLike<TransportMakeRequestResponse>;
+
+export const DEFAULT_TRANSPORT_BUFFER_SIZE = 30;
+
+/**
+ * Creates a `NewTransport`
+ *
+ * @param options
+ * @param makeRequest
+ */
+export function createTransport(
+  options: InternalBaseTransportOptions,
+  makeRequest: MakeTransportRequest,
+  buffer: PromiseBuffer<TransportResponse> = makePromiseBuffer(options.bufferSize || DEFAULT_TRANSPORT_BUFFER_SIZE),
+): NewTransport {
+  let rateLimits: RateLimits = {};
+
+  const flush = (timeout?: number): PromiseLike<boolean> => buffer.drain(timeout);
+
+  function send(envelope: Envelope, category: TransportCategory): PromiseLike<TransportResponse> {
+    const request: TransportRequest = {
+      category,
+      body: serializeEnvelope(envelope),
+    };
+
+    // Don't add to buffer if transport is already rate-limited
+    if (isRateLimited(rateLimits, category)) {
+      return rejectedSyncPromise({
+        status: 'rate_limit',
+        reason: getRateLimitReason(rateLimits, category),
+      });
+    }
+
+    const requestTask = (): PromiseLike<TransportResponse> =>
+      makeRequest(request).then(({ body, headers, reason, statusCode }): PromiseLike<TransportResponse> => {
+        const status = eventStatusFromHttpCode(statusCode);
+        if (headers) {
+          rateLimits = updateRateLimits(rateLimits, headers);
+        }
+        if (status === 'success') {
+          return resolvedSyncPromise({ status, reason });
+        }
+        return rejectedSyncPromise({
+          status,
+          reason:
+            reason || body || status === 'rate_limit'
+              ? getRateLimitReason(rateLimits, category)
+              : 'Unknown transport error',
```

Comment:
There might be a bug hidden here. When `reason` and  `body` are both undefined and `status` is `'invalid'` for example, the ternary operator will evaluate `getRateLimit(...)` and include a message "To many requests...", which is not right in that case.

IMO we should split this operator up to make it a bit more clear what's exactly happening here.

Reply:
I put it like this so that the execution is lazy (and is as bundle size efficient as possible). Fixed the bug you talked about, nice catch!